File: 2024/04 April/Remove every kth node/main.py
# Link: https://www.geeksforgeeks.org/problems/remove-every-kth-node/1


class node:

    def __init__(self, x):
        self.data = x
        self.next = None


def createLinkedList(arr):
    head = node(arr[0])
    curr = head
    for i in range(1, len(arr)):
        new_node = node(arr[i])
        curr.next = new_node
        curr = curr.next

    return head


def printlist(ptr):
    while ptr != None:
        print(ptr.data, end=" ")
        ptr = ptr.next
    print()


class Solution:
    def deleteK(self, head, k):
        if k == 1:
            return None
        node = head
        while node != None:
            for i in range(k - 2):
                if node == None:
                    return head
                node = node.next
            if node == None:
                return head
            if node.next != None:
                node.next = node.next.next
            node = node.next

        return head


if __name__ == "__main__":
    obj = Solution()
    arr = [1, 2, 3, 4, 5, 6, 7, 8]
    k = 2
    head = createLinkedList(arr)
    head = obj.deleteK(head, k)
    printlist(head)

    arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    k = 3
    head = createLinkedList(arr)
    head = obj.deleteK(head, k)
    printlist(head)

File: 2024/04 April/Remove every kth node/main.cpp
// Link: https://www.geeksforgeeks.org/problems/remove-every-kth-node/1

#include <bits/stdc++.h>
using namespace std;

struct Node
{
    int data;
    struct Node *next;

    Node(int x)
    {
        data = x;
        next = NULL;
    }
};

void append(struct Node **head_ref, struct Node **tail_ref, int new_data)
{
    struct Node *new_node = new Node(new_data);

    if (*head_ref == NULL)
        *head_ref = new_node;
    else
        (*tail_ref)->next = new_node;

    *tail_ref = new_node;
}

class Solution
{
public:
    Node *deleteK(Node *head, int K)
    {
        if (K == 1)
            return NULL;
        Node *curr = head;
        int cnt = 1;
        while (curr != NULL)
        {
            cnt++;
            if (cnt == K && curr->next != NULL)
            {
                curr->next = curr->next->next;
                cnt = 1;
            }
            curr = curr->next;
        }
        return head;
    }
};

int main()
{
    Solution obj;
    vector<int> arr = {1, 2, 3, 4, 5, 6, 7, 8};
    int k = 2;
    Node *head = NULL, *tail = NULL;
    for (int i = 0; i < arr.size(); i++)
        append(&head, &tail, arr[i]);
    Node *res = obj.deleteK(head, k);
    while (res != NULL)
    {
        cout << res->data << " ";
        res = res->next;
    }
    cout << endl;

    arr = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    k = 3;
    head = NULL, tail = NULL;
    for (int i = 0; i < arr.size(); i++)
        append(&head, &tail, arr[i]);
    res = obj.deleteK(head, k);
    while (res != NULL)
    {
        cout << res->data << " ";
        res = res->next;
    }
    cout << endl;
    return 0;
}